interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,394 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const customDeviceSummary = `provides access to custom devices specified via the gadget snap`
+
+const customDeviceBaseDeclarationSlots = `
+  custom-device:
+    allow-installation:
+      slot-snap-type:
+        - gadget
+    allow-connection:
+      plug-attributes:
+        content: $SLOT(custom-device)
+    deny-auto-connection: true
+`
+
+var (
+	// A cryptic, uninformative error message that we use only on impossible code paths
+	customDeviceInternalError = errors.New(`custom-device interface internal error`)
+
+	// Validating regexp for filesystem paths
+	customDevicePathRegexp = regexp.MustCompile(`^/[^"@]*$`)
+
+	// Validating regexp for udev device names.
+	// We forbid:
+	// - `|`: it's valid for udev, but more work for us
+	// - `{}`: have a special meaning for AppArmor
+	// - `"`: it's just dangerous (both for udev and AppArmor)
+	// - `\`: also dangerous
+	customDeviceUDevDeviceRegexp = regexp.MustCompile(`^/dev/[^"|{}\\]+$`)
+
+	// Validating regexp for udev tag values (all but kernel devices)
+	customDeviceUDevValueRegexp = regexp.MustCompile(`^[^"{}\\]+$`)
+)
+
+// customDeviceInterface allows sharing customDevice between snaps
+type customDeviceInterface struct{}
+
+func (iface *customDeviceInterface) validateFilePath(path string, attrName string) error {
+	if !customDevicePathRegexp.MatchString(path) {
+		return fmt.Errorf(`custom-device %q path must start with / and cannot contain special characters: %q`, attrName, path)
+	}
+
+	if !cleanSubPath(path) {
+		return fmt.Errorf(`custom-device %q path is not clean: %q`, attrName, path)
+	}
+
+	if _, err := utils.NewPathPattern(path); err != nil {
+		return fmt.Errorf(`custom-device %q path cannot be used: %v`, attrName, err)
+	}
+
+	// We don't allow "**" (a single "*" is already transformed into a "**"
+	// apparmor rule anyway)
```

Comment:
where does this transformation actually happen? I don't see where this happens, AFAICT if we include a "*" in devices, then a single "*" will continue to show up in the AppArmorSpecification snippet

Reply:
Did this just get removed in https://github.com/snapcore/snapd/pull/11272/commits/20cb846b6f0a32fd0fd09434705f8b28911f5d97 ? Does that comment need to be updated then?

Reply:
ah yes that makes more sense

Reply:
Oops, indeed I forgot to update the comment. Done now :-)